kiali/kiali
# Autodetect gateways across namespaces

### Describe the change

If a user specifies `istio.component_status`, then that will be used exclusively for the istio status checks. If a user doesn't specify `istio.component_status`, then Kiali will look for gateway workloads across all namespaces. Since we look for gateway workloads across namespaces, this PR also removes the gateway related Kiali config options. It also moves the logic for determining if a workload is a gateway or not to the backend whereas before it was duplicated in both the frontend and the backend. This PR also changes ingress gateways to be "not core" in the istio status health checks. Now when you deploy Kiali without creating an ingress/egress gateway, it will no longer complain at you in the masthead.

### Steps to test the PR

#### Test gateway autodetection
1. Deploy Istio + Kiali
2. Create a gateway with gateway-injection:
    ```
    kubectl create namespace istio-ingress
    helm install istio-ingressgateway istio/gateway -n istio-ingress
    ```
3. Ensure Kiali doesn't have a "missing sidecar" badge for the gateway workload.
4. Create a gateway with gateway-api
    ```
    kubectl create namespace istio-ingress
    kubectl apply -f - <<EOF
    apiVersion: gateway.networking.k8s.io/v1
    kind: Gateway
    metadata:
      name: gateway
      namespace: istio-ingress
    spec:
      gatewayClassName: istio
      listeners:
      - name: default
        hostname: "*.example.com"
        port: 80
        protocol: HTTP
        allowedRoutes:
          namespaces:
            from: All
    EOF
    ```
5. Ensure Kiali doesn't have a "missing sidecar" badge for the gateway workload.

#### Test specifying `component_status`
1. Deploy Istio + Kiali
2. Set istio component status to something wrong:
    ```
      istio:
        component_status
          components:
          - app_label: nonexistantgateway
            is_core: false
            is_proxy: true
            namespace: doesnotexist
    ```
3. Masthead should show a warning just for what you've set.

#### Test deploying Istio without gateways doesn't show warning

1. Deploy Sail + Kiali. Use this PR to easily deploy Kiali + sail: https://github.com/kiali/kiali/pull/7876. Then rebuild/reload kiali with this PR.
2. Ensure no warnings in masthead

### Automation testing

Included unit tests.

### Issue reference

#7882
